File: python/dp/fibonacci.py
# Copyright [2018] [Haibo Yan]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np

class Fibonnaci:
    def __init__(self):
        a = [[1, 1], [1, 0]]
        self._cache = {1:np.array(a)}
        k = 1
        for i in range(2, 8):
            n = self._cache[k].dot(self._cache[k])
            k *= 2
            print(k)
            self._cache[k] = n

    def nth(self, n):
        if n == 1 or n == 2:
            return 1
        n -= 1
        v = n & -n
        m = self._cache[v]
        n -= v

        while n != 0:
            v = n & -n
            m = m.dot(self._cache[v])
            n -= v

        return m[0][0]

File: c++/algorithm/dp/Fibonacci.cpp
//Copyright [2018] [Haibo Yan]
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

//
// Created by haibo on 3/16/18.
//

#include "Fibonacci.h"
#include <iostream>

vector<vector<long>> Fibonacci::times(vector<vector<long>> m1, vector<vector<long>> m2, int r){
    vector<vector<long>> p;

    for (int i = 0; i < r; i++) {
        vector<long> l;
        for (int j = 0; j < r; j++) {
            l.push_back(0);
        }
        p.push_back(l);
    }

    for (int i = 0; i < r; i++) {
        for (int j = 0; j < r; j++) {
            for (int k = 0; k < r; k++) {
                p[i][j] += m1[i][k] * m2[k][j];
            }
        }
    }

    return p;
}

Fibonacci::Fibonacci() {
    vector<vector<long>> one = {
            {1, 1},
            {1, 0}
    };

    _twoFi[1] = one;

    for (int i = 1; i < 8; i++) {
        int k = 1 << i;
        vector<vector<long>> m = _twoFi[k / 2];
        _twoFi[k] = times(m, m, 2);
    }
}

long Fibonacci::nth(int n) {
    if (n < 3) {
        return 1;
    }
    n--;

    int l = n & -n;
    vector<vector<long>> m = _twoFi[l];
    n -= l;

    while(n > 0) {
        l = n & -n;
        m = times(m, _twoFi[l], 2);
        n -= l;
    }

    return m[0][0];
}